### PR TITLE
FFM blocks simplified (stepping stone for caching & SIMD)

### PR DESCRIPTION
**Pull request by ggaspersic**
Overview:

* removal of unused (old) related functionality
* replacing weights: Vec<WeightAndOptimizerData<L>>, with 2 separate vectors (this will allow us later on to have easier SIMD & caching (when doing forward pass)). The impact of this is that the reading also needed to be changed.
* removed the redundant usage of macros
* updated variable names to make them more descriptive

### CODE REVIEW

File: src/block_ffm.rs [hunk 1]
```diff
@@ -225,149 +127,166 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                            feature_num += 1;
                         }
+                    }
+
+                    let mut ffm_values_offset = 0;
+                    for (i, feature) in fb.ffm_buffer.iter().enumerate() {
+                        let feature_value = feature.value;
+                        let feature_index = feature.hash as usize;
+                        let feature_contra_field_index = feature.contra_field_index as usize;
+
```
Comment (SkBlaz) [inline src/block_ffm.rs:215]: Not needed spaces

File: src/block_ffm.rs [hunk 2]
```diff
@@ -398,177 +317,228 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                ),
+                _MM_HINT_T0,
+            );
+
+            /* We first prepare "contra_fields" or collapsed field embeddings, where we sum all individual feature embeddings
+               We need to be careful to:
+               - handle fields with zero features present
+               - handle values on diagonal - we want to be able to exclude self-interactions later (we pre-substract from wsum)
+               - optimize for just copying the embedding over when looking at first feature of the field, and add embeddings for the rest
+               - optimize for very common case of value of the feature being 1.0 - avoid multiplications
```
Comment (SkBlaz) [inline src/block_ffm.rs:333]: Thinking about this, case of value being not present might also make sense to separate out of multiplying redundantly (many such ones)

File: src/block_ffm.rs [hunk 3]
```diff
@@ -398,177 +317,228 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+
+                    ffm_buffer_index += 1;
+                    feature_num += 1;
+                }
+            }
+
+            let mut f1_offset = 0;
+            let mut f1_index_offset = 0;
+            let mut f1_ffmk = 0;
+            let mut diagonal_row = 0;
```
Comment (SkBlaz) [inline src/block_ffm.rs:411]: Name is a bit off. `row` implies some kind of vector imo, this is just a scalar

**Comment by SkBlaz** [comment]
General: Changing weight structure might (or might not) have implications for reading speed/some intermediary operations; assuming this is negligible speed wise? (there could be a mini test/benchmark just so we're sure, even though it's most likely not an issue)

File: src/block_ffm.rs (hunk 1 above)
Comment (ggaspersic) [reply src/block_ffm.rs:215]: Tbh I actually want to reach a point where we just do `cargo fmt`.

**Review by bbenshalom — APPROVED** [review]

**Review by SkBlaz — APPROVED** [review]

File: src/block_ffm.rs [hunk 4]
```diff
@@ -226,148 +127,165 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
                                     }
-                                });
-                                ffm_buffer_index += 1;
-                                feature_num += 1;
+
+                                    offset += fc;
+                                    feature_index += ffmk_as_usize;
+                                }
+                            } else {
+                                for _z in 0..ffm_fields_count_as_usize {
```
Comment (SkBlaz) [inline src/block_ffm.rs:194]: I think the convention is `_` instead of `_z` but not really relevant

File: src/block_ffm.rs [hunk 5]
```diff
@@ -226,148 +127,165 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
 
-                        let mut ffm_values_offset = 0;
-                        for (i, left_hash) in fb.ffm_buffer.iter().enumerate() {
-                            let contra_offset = (left_hash.contra_field_index * fb.ffm_fields_count) as usize;
-                            let mut vv = 0;
-                            let contra_offset2 = contra_offset / FFMK as usize;
-                            let left_hash_value = left_hash.value;
-                            let left_hash_contra_field_index = left_hash.contra_field_index;
-                            let left_hash_hash = left_hash.hash as usize;
-                            //let LEFT_HASH_VALUE = left_hash_value;
```
Comment (SkBlaz) [inline src/block_ffm.rs:304]: Is this comment redundant?

Comment (ggaspersic) [reply src/block_ffm.rs:304]: This was removed right?

File: src/block_ffm.rs [hunk 6]
```diff
@@ -398,177 +316,228 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                    ffm_buffer_index += 1;
+                    feature_num += 1;
+                }
+            }
+
+            let mut f1_offset = 0;
+            let mut f1_index_offset = 0;
+            let mut f1_ffmk = 0;
+            let mut self_interaction_index = 0;
+            for f1 in 0..ffm_fields_count_as_usize {
```
Comment (SkBlaz) [inline src/block_ffm.rs:411]: `f1` and `f2` seem odd variable names, but ok

Comment (ggaspersic) [reply src/block_ffm.rs:411]: I agree, but I left them the same as before

File: src/block_ffm.rs (hunk 4 above)
Comment (ggaspersic) [reply src/block_ffm.rs:194]: I think you added that in the previous commit to `main` :)

**Pull request merged**